### DISCUSSION TITLE
Validating gnupg signatures

### DISCUSSION THREAD

**Discussion by karlbrown-va** [Homebrew/discussions#4113] [Everyday usage]
### Output of `brew config`

```shell
n/a
```

### Output of `brew doctor`

```shell
n/a
```

### Description of issue

Hi, I'm interested in verifying the signatures on homebrew gnupg installs.

I'm trying to figure out the best way to do this. I came across this discussion to add this feature to Homebrew but it seems it's not something planned currently.
https://github.com/Homebrew/brew/issues/5852

This is my current thinking:
1) If no gnupg installed, you can manually verify the SHA256 sum here before installing:
`brew info --json gpg | jq -r ".[].urls.stable"`
and compare to SHA256 sum here:
https://gnupg.org/download/integrity_check.html

If you also wanted to verify signatures, you'd need some other trusted PGP software to do so.

2) If a (trusted) gnupg is already installed, before installing the new version, you could do the SHA256 check as above. However if I want to verify signatures, gnups's published signatures reference the tarball, but as I understand it the tarball then gets turned into a bottle.

How can I be sure that the exact tarball ends up on my disk? (example -- if unlikely -- threat: attacker poisons homebrew package directory). I'm not sure about how the bottling process works, but the result is the bottle has a different checksum than the gnupg tarball. Any thoughts?

**Comment by SMillerDev** [comment]
If an attacker compromises Homebrews GitHub there is nothing that you can use to verify since all Homebrew information comes from there. If this is a likely attack vector for you, you might be better off without a package manager.

**Reply by karlbrown-va** [reply]
Thanks - yes I understand there are issues w.r.t. trust with a package manager.

What about an alternative like the following:
1) Download manually the gnupg tarball and confirm signatures
2) Use this tarball to build a bottle locally
3) Install the (locally-built) bottle

Is this possible? If so how would it be done?

Also could you point me to where the gnupg bottles are built? I was poking around in github actions but wasn't able to find it. I assumed it's somewhere here but I didn't find it:
https://github.com/Homebrew/homebrew-core/pull/120045/checks

**Reply by SMillerDev** [reply]
If you don't want to trust the package manager wouldn't prebuilt binaries from upstream be easier?

**Reply by karlbrown-va** [reply]
Ok, I think I found a solution that is somewhat reasonable for the quasi-paranoid:
1) Manually download the gnu tarball from the location specified in `brew info --json gpg | jq -r ".[].urls.stable.url"`
2) Confirm sha256 sum matches what the homebrew formula claims, and the actual tarball
```
brew info --json gpg | jq -r ".[].urls.stable.checksum"
shasum -a 256 ~/Downloads/gnupg-2.4.0.tar.bz2
```
3) Check signatures on tarball
`gpg --verify ~/Downloads/gnupg-2.4.0.tar.bz2.sig ~/Downloads/gnupg-2.4.0.tar.bz2`
3) Build locally from source with `brew install -s gnupg` (This will download the same tarball and then build it locally, though it will pull bottled dependencies I think). Nonetheless at least the core gnupg code is verified as signed with this approach.

The main threats I think this would defend against would be a compromised gnupg repository (the signature should fail, assuming you've only trusted the correct signatures), or a compromised brew package for gnupg. It would not help if dependencies were compromised however.

If you want to build entirely from source w/o relying on anything from brew, you can follow these instructions:
https://carreno.me/articles/how-to-build-gnupg-2-from-source-on-os-x

Finally an alternative is to install the [GnuPG](https://gpgtools.org/), which has its own signatures, but now you're trusting them... (or [GPGOsx](https://sourceforge.net/p/gpgosx/docu/Download/))

(Edit as I hadn't seen your comment - re: pre-built binaries, I don't think gnupg project publishes osx binaries)

**Reply by SMillerDev** [reply]
https://www.gnupg.org/download/index.html seems to suggest they do.

**Reply by karlbrown-va** [reply]
I think those OSX binaries are made by different teams (which I linked to above - GnuPG and GPGOSx)

**Reply by SMillerDev** [reply]
They're trusted by the official project to be referred from their official site though. Unless you don't trust GnuPG to define official packagers, but that's a very big hole you'd be digging for yourself.

**Reply by karlbrown-va** [reply]
Yeah, it's a fair point. They likely trust those packagers, but it does add another of trust required, as they are one step removed from the core maintainers. I think the approach I outlined above allows me to stay in the brew system, but provides some additional verification.
